From 2e6fcdfebbc3ea2875712b526809234ab728e56e Mon Sep 17 00:00:00 2001
From: Michal Gorny <mgorny@gentoo.org>
Date: Fri, 4 Oct 2019 20:30:02 +0000
Subject: [PATCH] [clang-tools-extra] [cmake] Link against libclang-cpp
 whenever possible

Use clang_target_link_libraries() in order to support linking against
libclang-cpp instead of static libraries.

Differential Revision: https://reviews.llvm.org/D68448

git-svn-id: https://llvm.org/svn/llvm-project/clang-tools-extra/trunk@373786 91177308-0d34-0410-b5e6-96231b3b80d8
---
 clang-apply-replacements/tool/CMakeLists.txt          |  7 +++++--
 clang-change-namespace/tool/CMakeLists.txt            |  7 +++++--
 clang-doc/tool/CMakeLists.txt                         |  7 +++++--
 .../find-all-symbols/tool/CMakeLists.txt              |  5 ++++-
 clang-include-fixer/tool/CMakeLists.txt               |  7 +++++--
 clang-move/tool/CMakeLists.txt                        |  7 +++++--
 clang-query/tool/CMakeLists.txt                       |  7 +++++--
 clang-reorder-fields/tool/CMakeLists.txt              |  7 +++++--
 clang-tidy/CMakeLists.txt                             |  2 +-
 clang-tidy/tool/CMakeLists.txt                        |  7 +++++--
 clangd/fuzzer/CMakeLists.txt                          |  7 +++++--
 clangd/index/dex/dexp/CMakeLists.txt                  |  5 ++++-
 clangd/indexer/CMakeLists.txt                         |  7 +++++--
 clangd/tool/CMakeLists.txt                            |  9 ++++++---
 clangd/unittests/CMakeLists.txt                       |  9 ++++++---
 clangd/xpc/test-client/CMakeLists.txt                 |  7 +++++--
 modularize/CMakeLists.txt                             |  2 +-
 pp-trace/CMakeLists.txt                               |  2 +-
 tool-template/CMakeLists.txt                          |  2 +-
 unittests/clang-apply-replacements/CMakeLists.txt     |  7 +++++--
 unittests/clang-change-namespace/CMakeLists.txt       |  7 +++++--
 unittests/clang-doc/CMakeLists.txt                    |  7 +++++--
 unittests/clang-include-fixer/CMakeLists.txt          |  7 +++++--
 .../find-all-symbols/CMakeLists.txt                   |  5 ++++-
 unittests/clang-move/CMakeLists.txt                   |  7 +++++--
 unittests/clang-query/CMakeLists.txt                  |  7 +++++--
 unittests/clang-tidy/CMakeLists.txt                   | 11 +++++++----
 27 files changed, 120 insertions(+), 51 deletions(-)

diff --git a/clang-apply-replacements/tool/CMakeLists.txt b/clang-apply-replacements/tool/CMakeLists.txt
index d15a8b1af..1ed734c11 100644
--- a/clang-apply-replacements/tool/CMakeLists.txt
+++ b/clang-apply-replacements/tool/CMakeLists.txt
@@ -5,12 +5,15 @@ set(LLVM_LINK_COMPONENTS
 add_clang_tool(clang-apply-replacements
   ClangApplyReplacementsMain.cpp
   )
-target_link_libraries(clang-apply-replacements
+clang_target_link_libraries(clang-apply-replacements
   PRIVATE
-  clangApplyReplacements
   clangBasic
   clangFormat
   clangRewrite
   clangToolingCore
   clangToolingRefactoring
   )
+target_link_libraries(clang-apply-replacements
+  PRIVATE
+  clangApplyReplacements
+  )
diff --git a/clang-change-namespace/tool/CMakeLists.txt b/clang-change-namespace/tool/CMakeLists.txt
index 702bad368..ae48a5e0f 100644
--- a/clang-change-namespace/tool/CMakeLists.txt
+++ b/clang-change-namespace/tool/CMakeLists.txt
@@ -7,12 +7,11 @@ set(LLVM_LINK_COMPONENTS
 add_clang_tool(clang-change-namespace
   ClangChangeNamespace.cpp
   )
-target_link_libraries(clang-change-namespace
+clang_target_link_libraries(clang-change-namespace
   PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
-  clangChangeNamespace
   clangFormat
   clangFrontend
   clangRewrite
@@ -20,3 +19,7 @@ target_link_libraries(clang-change-namespace
   clangTooling
   clangToolingCore
   )
+target_link_libraries(clang-change-namespace
+  PRIVATE
+  clangChangeNamespace
+  )
diff --git a/clang-doc/tool/CMakeLists.txt b/clang-doc/tool/CMakeLists.txt
index de8c9bcbf..7e7147886 100644
--- a/clang-doc/tool/CMakeLists.txt
+++ b/clang-doc/tool/CMakeLists.txt
@@ -4,16 +4,19 @@ add_clang_tool(clang-doc
   ClangDocMain.cpp
   )
 
-target_link_libraries(clang-doc
+clang_target_link_libraries(clang-doc
   PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangFrontend
-  clangDoc
   clangTooling
   clangToolingCore
   )
+target_link_libraries(clang-doc
+  PRIVATE
+  clangDoc
+  )
 
 install(FILES ../assets/clang-doc-default-stylesheet.css
   DESTINATION share/clang
diff --git a/clang-include-fixer/find-all-symbols/tool/CMakeLists.txt b/clang-include-fixer/find-all-symbols/tool/CMakeLists.txt
index 7f101ebd9..8f5509d22 100644
--- a/clang-include-fixer/find-all-symbols/tool/CMakeLists.txt
+++ b/clang-include-fixer/find-all-symbols/tool/CMakeLists.txt
@@ -4,7 +4,7 @@ add_clang_tool(find-all-symbols
   FindAllSymbolsMain.cpp
   )
 
-target_link_libraries(find-all-symbols
+clang_target_link_libraries(find-all-symbols
   PRIVATE
   clangAST
   clangASTMatchers
@@ -13,6 +13,9 @@ target_link_libraries(find-all-symbols
   clangLex
   clangSerialization
   clangTooling
+  )
+target_link_libraries(find-all-symbols
+  PRIVATE
   findAllSymbols
   )
 
diff --git a/clang-include-fixer/tool/CMakeLists.txt b/clang-include-fixer/tool/CMakeLists.txt
index 5b600a463..3936ac1e8 100644
--- a/clang-include-fixer/tool/CMakeLists.txt
+++ b/clang-include-fixer/tool/CMakeLists.txt
@@ -4,16 +4,19 @@ add_clang_tool(clang-include-fixer
   ClangIncludeFixer.cpp
   )
 
-target_link_libraries(clang-include-fixer
+clang_target_link_libraries(clang-include-fixer
   PRIVATE
   clangBasic
   clangFormat
   clangFrontend
-  clangIncludeFixer
   clangRewrite
   clangSerialization
   clangTooling
   clangToolingCore
+  )
+target_link_libraries(clang-include-fixer
+  PRIVATE
+  clangIncludeFixer
   findAllSymbols
   )
 
diff --git a/clang-move/tool/CMakeLists.txt b/clang-move/tool/CMakeLists.txt
index b6051e4fa..a0c9c20d7 100644
--- a/clang-move/tool/CMakeLists.txt
+++ b/clang-move/tool/CMakeLists.txt
@@ -4,16 +4,19 @@ add_clang_tool(clang-move
   ClangMove.cpp
   )
 
-target_link_libraries(clang-move
+clang_target_link_libraries(clang-move
   PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangFormat
   clangFrontend
-  clangMove
   clangRewrite
   clangSerialization
   clangTooling
   clangToolingCore
   )
+target_link_libraries(clang-move
+  PRIVATE
+  clangMove
+  )
diff --git a/clang-query/tool/CMakeLists.txt b/clang-query/tool/CMakeLists.txt
index 7071c94cf..be83a7c08 100644
--- a/clang-query/tool/CMakeLists.txt
+++ b/clang-query/tool/CMakeLists.txt
@@ -3,14 +3,17 @@ include_directories(${CMAKE_CURRENT_SOURCE_DIR}/..)
 add_clang_tool(clang-query
   ClangQuery.cpp
   )
-target_link_libraries(clang-query
+clang_target_link_libraries(clang-query
   PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangDynamicASTMatchers
   clangFrontend
-  clangQuery
   clangSerialization
   clangTooling
   )
+target_link_libraries(clang-query
+  PRIVATE
+  clangQuery
+  )
diff --git a/clang-reorder-fields/tool/CMakeLists.txt b/clang-reorder-fields/tool/CMakeLists.txt
index 718ee960a..b414f4f4d 100644
--- a/clang-reorder-fields/tool/CMakeLists.txt
+++ b/clang-reorder-fields/tool/CMakeLists.txt
@@ -2,13 +2,16 @@ add_clang_tool(clang-reorder-fields
   ClangReorderFields.cpp
   )
 
-target_link_libraries(clang-reorder-fields
+clang_target_link_libraries(clang-reorder-fields
   PRIVATE
   clangBasic
   clangFrontend
-  clangReorderFields
   clangRewrite
   clangSerialization
   clangTooling
   clangToolingCore
   )
+target_link_libraries(clang-reorder-fields
+  PRIVATE
+  clangReorderFields
+  )
diff --git a/clang-tidy/CMakeLists.txt b/clang-tidy/CMakeLists.txt
index 6dadb2717..8e747b32e 100644
--- a/clang-tidy/CMakeLists.txt
+++ b/clang-tidy/CMakeLists.txt
@@ -31,7 +31,7 @@ add_clang_library(clangTidy
   )
 
 if(CLANG_ENABLE_STATIC_ANALYZER)
-  target_link_libraries(clangTidy PRIVATE
+  clang_target_link_libraries(clangTidy PRIVATE
     clangStaticAnalyzerCore
     clangStaticAnalyzerFrontend
   )
diff --git a/clang-tidy/tool/CMakeLists.txt b/clang-tidy/tool/CMakeLists.txt
index fc2b4ebd3..073749a7d 100644
--- a/clang-tidy/tool/CMakeLists.txt
+++ b/clang-tidy/tool/CMakeLists.txt
@@ -11,14 +11,17 @@ add_clang_tool(clang-tidy
 add_dependencies(clang-tidy
   clang-resource-headers
   )
-target_link_libraries(clang-tidy
+clang_target_link_libraries(clang-tidy
   PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
-  clangTidy
   clangTooling
   clangToolingCore
+  )
+target_link_libraries(clang-tidy
+  PRIVATE
+  clangTidy
   ${ALL_CLANG_TIDY_CHECKS}
   )
 
diff --git a/clangd/fuzzer/CMakeLists.txt b/clangd/fuzzer/CMakeLists.txt
index 28191a3f5..90379822a 100644
--- a/clangd/fuzzer/CMakeLists.txt
+++ b/clangd/fuzzer/CMakeLists.txt
@@ -12,13 +12,16 @@ add_llvm_fuzzer(clangd-fuzzer
   DUMMY_MAIN DummyClangdMain.cpp
   )
 
-target_link_libraries(clangd-fuzzer
+clang_target_link_libraries(clangd-fuzzer
   PRIVATE
   clangBasic
-  clangDaemon
   clangFormat
   clangFrontend
   clangSema
   clangTooling
   clangToolingCore
   )
+target_link_libraries(clangd-fuzzer
+  PRIVATE
+  clangDaemon
+  )
diff --git a/clangd/index/dex/dexp/CMakeLists.txt b/clangd/index/dex/dexp/CMakeLists.txt
index ece339d70..a4edbb372 100644
--- a/clangd/index/dex/dexp/CMakeLists.txt
+++ b/clangd/index/dex/dexp/CMakeLists.txt
@@ -9,8 +9,11 @@ add_clang_executable(dexp
   Dexp.cpp
   )
 
-target_link_libraries(dexp
+clang_target_link_libraries(dexp
   PRIVATE
   clangBasic
+  )
+target_link_libraries(dexp
+  PRIVATE
   clangDaemon
   )
diff --git a/clangd/indexer/CMakeLists.txt b/clangd/indexer/CMakeLists.txt
index 92aae0643..edbced141 100644
--- a/clangd/indexer/CMakeLists.txt
+++ b/clangd/indexer/CMakeLists.txt
@@ -8,13 +8,16 @@ add_clang_executable(clangd-indexer
   IndexerMain.cpp
   )
 
-target_link_libraries(clangd-indexer
+clang_target_link_libraries(clangd-indexer
   PRIVATE
   clangAST
   clangBasic
-  clangDaemon
   clangFrontend
   clangIndex
   clangLex
   clangTooling
 )
+target_link_libraries(clangd-indexer
+  PRIVATE
+  clangDaemon
+)
diff --git a/clangd/tool/CMakeLists.txt b/clangd/tool/CMakeLists.txt
index 085ede378..9ebbe5f04 100644
--- a/clangd/tool/CMakeLists.txt
+++ b/clangd/tool/CMakeLists.txt
@@ -15,12 +15,10 @@ if(CLANGD_BUILD_XPC)
   list(APPEND CLANGD_XPC_LIBS "clangdXpcJsonConversions" "clangdXpcTransport")
 endif()
 
-target_link_libraries(clangd
+clang_target_link_libraries(clangd
   PRIVATE
   clangAST
   clangBasic
-  clangTidy
-  clangDaemon
   clangFormat
   clangFrontend
   clangSema
@@ -28,5 +26,10 @@ target_link_libraries(clangd
   clangToolingCore
   clangToolingRefactoring
   clangToolingSyntax
+  )
+target_link_libraries(clangd
+  PRIVATE
+  clangTidy
+  clangDaemon
   ${CLANGD_XPC_LIBS}
   )
diff --git a/clangd/unittests/CMakeLists.txt b/clangd/unittests/CMakeLists.txt
index d25745f94..7e298b6ad 100644
--- a/clangd/unittests/CMakeLists.txt
+++ b/clangd/unittests/CMakeLists.txt
@@ -78,23 +78,26 @@ add_unittest(ClangdUnitTests ClangdTests
   $<TARGET_OBJECTS:obj.clangDaemonTweaks>
   )
 
-target_link_libraries(ClangdTests
+clang_target_link_libraries(ClangdTests
   PRIVATE
   clangAST
   clangBasic
-  clangDaemon
   clangFormat
   clangFrontend
   clangIndex
   clangLex
   clangSema
   clangSerialization
-  clangTidy
   clangTooling
   clangToolingCore
   clangToolingInclusions
   clangToolingRefactoring
   clangToolingSyntax
+  )
+target_link_libraries(ClangdTests
+  PRIVATE
+  clangDaemon
+  clangTidy
   LLVMSupport
   LLVMTestingSupport
   )
diff --git a/clangd/xpc/test-client/CMakeLists.txt b/clangd/xpc/test-client/CMakeLists.txt
index 283599ecc..1bf01c63d 100644
--- a/clangd/xpc/test-client/CMakeLists.txt
+++ b/clangd/xpc/test-client/CMakeLists.txt
@@ -13,14 +13,17 @@ set(LLVM_LINK_COMPONENTS
     support
 )
 
-target_link_libraries(clangd-xpc-test-client
+clang_target_link_libraries(clangd-xpc-test-client
   PRIVATE
   clangBasic
-  clangDaemon
   clangFormat
   clangFrontend
   clangSema
   clangTooling
   clangToolingCore
+)
+target_link_libraries(clangd-xpc-test-client
+  PRIVATE
+  clangDaemon
   clangdXpcJsonConversions
 )
diff --git a/modularize/CMakeLists.txt b/modularize/CMakeLists.txt
index fa2c0e534..4caae81c4 100644
--- a/modularize/CMakeLists.txt
+++ b/modularize/CMakeLists.txt
@@ -11,7 +11,7 @@ add_clang_tool(modularize
   PreprocessorTracker.cpp
   )
 
-target_link_libraries(modularize
+clang_target_link_libraries(modularize
   PRIVATE
   clangAST
   clangBasic
diff --git a/pp-trace/CMakeLists.txt b/pp-trace/CMakeLists.txt
index 11b45ac63..be1d9715c 100644
--- a/pp-trace/CMakeLists.txt
+++ b/pp-trace/CMakeLists.txt
@@ -7,7 +7,7 @@ add_clang_tool(pp-trace
   PPCallbacksTracker.cpp
   )
 
-target_link_libraries(pp-trace
+clang_target_link_libraries(pp-trace
   PRIVATE
   clangAST
   clangBasic
diff --git a/tool-template/CMakeLists.txt b/tool-template/CMakeLists.txt
index 9a304d434..959bd3d1a 100644
--- a/tool-template/CMakeLists.txt
+++ b/tool-template/CMakeLists.txt
@@ -6,7 +6,7 @@ add_clang_executable(tool-template
   ToolTemplate.cpp
   )
 
-target_link_libraries(tool-template
+clang_target_link_libraries(tool-template
   PRIVATE
   clangAST
   clangASTMatchers
diff --git a/unittests/clang-apply-replacements/CMakeLists.txt b/unittests/clang-apply-replacements/CMakeLists.txt
index d3200d76b..b345527fe 100644
--- a/unittests/clang-apply-replacements/CMakeLists.txt
+++ b/unittests/clang-apply-replacements/CMakeLists.txt
@@ -11,10 +11,13 @@ add_extra_unittest(ClangApplyReplacementsTests
   ApplyReplacementsTest.cpp
   )
 
-target_link_libraries(ClangApplyReplacementsTests
+clang_target_link_libraries(ClangApplyReplacementsTests
   PRIVATE
-  clangApplyReplacements
   clangBasic
   clangToolingCore
   clangToolingRefactoring
   )
+target_link_libraries(ClangApplyReplacementsTests
+  PRIVATE
+  clangApplyReplacements
+  )
diff --git a/unittests/clang-change-namespace/CMakeLists.txt b/unittests/clang-change-namespace/CMakeLists.txt
index 9c949723f..d66f85da4 100644
--- a/unittests/clang-change-namespace/CMakeLists.txt
+++ b/unittests/clang-change-namespace/CMakeLists.txt
@@ -15,12 +15,11 @@ add_extra_unittest(ClangChangeNamespaceTests
   ChangeNamespaceTests.cpp
   )
 
-target_link_libraries(ClangChangeNamespaceTests
+clang_target_link_libraries(ClangChangeNamespaceTests
   PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
-  clangChangeNamespace
   clangFormat
   clangFrontend
   clangRewrite
@@ -28,3 +27,7 @@ target_link_libraries(ClangChangeNamespaceTests
   clangTooling
   clangToolingCore
   )
+target_link_libraries(ClangChangeNamespaceTests
+  PRIVATE
+  clangChangeNamespace
+  )
diff --git a/unittests/clang-doc/CMakeLists.txt b/unittests/clang-doc/CMakeLists.txt
index 292a1d7a5..7934cd17c 100644
--- a/unittests/clang-doc/CMakeLists.txt
+++ b/unittests/clang-doc/CMakeLists.txt
@@ -20,12 +20,11 @@ add_extra_unittest(ClangDocTests
   YAMLGeneratorTest.cpp
   )
 
-target_link_libraries(ClangDocTests
+clang_target_link_libraries(ClangDocTests
   PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
-  clangDoc
   clangFormat
   clangFrontend
   clangRewrite
@@ -33,3 +32,7 @@ target_link_libraries(ClangDocTests
   clangTooling
   clangToolingCore
   )
+target_link_libraries(ClangDocTests
+  PRIVATE
+  clangDoc
+  )
diff --git a/unittests/clang-include-fixer/CMakeLists.txt b/unittests/clang-include-fixer/CMakeLists.txt
index 997aa1459..0c0954c2a 100644
--- a/unittests/clang-include-fixer/CMakeLists.txt
+++ b/unittests/clang-include-fixer/CMakeLists.txt
@@ -16,16 +16,19 @@ add_extra_unittest(ClangIncludeFixerTests
   FuzzySymbolIndexTests.cpp
   )
 
-target_link_libraries(ClangIncludeFixerTests
+clang_target_link_libraries(ClangIncludeFixerTests
   PRIVATE
   clangBasic
   clangFormat
   clangFrontend
-  clangIncludeFixer
   clangRewrite
   clangSerialization
   clangTooling
   clangToolingCore
+  )
+target_link_libraries(ClangIncludeFixerTests
+  PRIVATE
+  clangIncludeFixer
   findAllSymbols
   )
 
diff --git a/unittests/clang-include-fixer/find-all-symbols/CMakeLists.txt b/unittests/clang-include-fixer/find-all-symbols/CMakeLists.txt
index 427aa8ed8..828d4347a 100644
--- a/unittests/clang-include-fixer/find-all-symbols/CMakeLists.txt
+++ b/unittests/clang-include-fixer/find-all-symbols/CMakeLists.txt
@@ -12,7 +12,7 @@ add_extra_unittest(FindAllSymbolsTests
   FindAllSymbolsTests.cpp
   )
 
-target_link_libraries(FindAllSymbolsTests
+clang_target_link_libraries(FindAllSymbolsTests
   PRIVATE
   clangAST
   clangASTMatchers
@@ -21,5 +21,8 @@ target_link_libraries(FindAllSymbolsTests
   clangLex
   clangSerialization
   clangTooling
+  )
+target_link_libraries(FindAllSymbolsTests
+  PRIVATE
   findAllSymbols
   )
diff --git a/unittests/clang-move/CMakeLists.txt b/unittests/clang-move/CMakeLists.txt
index 1d5347fe3..468c65c36 100644
--- a/unittests/clang-move/CMakeLists.txt
+++ b/unittests/clang-move/CMakeLists.txt
@@ -15,16 +15,19 @@ add_extra_unittest(ClangMoveTests
   ClangMoveTests.cpp
   )
 
-target_link_libraries(ClangMoveTests
+clang_target_link_libraries(ClangMoveTests
   PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangFormat
   clangFrontend
-  clangMove
   clangRewrite
   clangSerialization
   clangTooling
   clangToolingCore
   )
+target_link_libraries(ClangMoveTests
+  PRIVATE
+  clangMove
+  )
diff --git a/unittests/clang-query/CMakeLists.txt b/unittests/clang-query/CMakeLists.txt
index 2177764c4..975664259 100644
--- a/unittests/clang-query/CMakeLists.txt
+++ b/unittests/clang-query/CMakeLists.txt
@@ -11,14 +11,17 @@ add_extra_unittest(ClangQueryTests
   QueryParserTest.cpp
   )
 
-target_link_libraries(ClangQueryTests
+clang_target_link_libraries(ClangQueryTests
   PRIVATE
   clangAST
   clangASTMatchers
   clangBasic
   clangDynamicASTMatchers
   clangFrontend
-  clangQuery
   clangSerialization
   clangTooling
   )
+target_link_libraries(ClangQueryTests
+  PRIVATE
+  clangQuery
+  )
diff --git a/unittests/clang-tidy/CMakeLists.txt b/unittests/clang-tidy/CMakeLists.txt
index 93b49f546..287b431c2 100644
--- a/unittests/clang-tidy/CMakeLists.txt
+++ b/unittests/clang-tidy/CMakeLists.txt
@@ -21,7 +21,7 @@ add_extra_unittest(ClangTidyTests
   TransformerClangTidyCheckTest.cpp
   )
 
-target_link_libraries(ClangTidyTests
+clang_target_link_libraries(ClangTidyTests
   PRIVATE
   clangAST
   clangASTMatchers
@@ -29,6 +29,12 @@ target_link_libraries(ClangTidyTests
   clangFrontend
   clangLex
   clangSerialization
+  clangTooling
+  clangToolingCore
+  clangToolingRefactoring
+  )
+target_link_libraries(ClangTidyTests
+  PRIVATE
   clangTidy
   clangTidyAndroidModule
   clangTidyGoogleModule
@@ -36,7 +42,4 @@ target_link_libraries(ClangTidyTests
   clangTidyObjCModule
   clangTidyReadabilityModule
   clangTidyUtils
-  clangTooling
-  clangToolingCore
-  clangToolingRefactoring
   )
-- 
2.25.0

